plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/transport/IngestActionFilter.java

```diff
@@ -25,65 +25,108 @@
-            processIndexRequest((IndexRequest) request, pipeline);
+            processIndexRequest(action, listener, chain, (IndexRequest) request, pipelineId);
         } else if (request instanceof BulkRequest) {
             BulkRequest bulkRequest = (BulkRequest) request;
-            List<ActionRequest> actionRequests = bulkRequest.requests();
-            for (ActionRequest actionRequest : actionRequests) {
-                if (actionRequest instanceof IndexRequest) {
-                    processIndexRequest((IndexRequest) actionRequest, pipeline);
-                }
-            }
+            processBulkIndexRequest(action, listener, chain, bulkRequest, pipelineId, bulkRequest.requests().iterator());
+        } else {
+            chain.proceed(action, request, listener);
         }
-        return true;
     }
 
-    // TODO: this should be delegated to a PipelineExecutor service that executes on a different thread (pipeline TP)
-    void processIndexRequest(IndexRequest indexRequest, Pipeline pipeline) {
+    @Override
+    public void apply(String action, ActionResponse response, ActionListener listener, ActionFilterChain chain) {
+        chain.proceed(action, response, listener);
+    }
+
+    void processIndexRequest(String action, ActionListener listener, ActionFilterChain chain, IndexRequest indexRequest, String pipelineId) {
         Map<String, Object> sourceAsMap = indexRequest.sourceAsMap();
         Data data = new Data(indexRequest.index(), indexRequest.type(), indexRequest.id(), sourceAsMap);
-        pipeline.execute(data);
-        indexRequest.source(data.getDocument());
+        executionService.execute(data, pipelineId, new PipelineExecutionService.Listener() {
+            @Override
+            public void executed(Data data) {
+                if (data.isModified()) {
+                    indexRequest.source(data.getDocument());
+                }
+                chain.proceed(action, indexRequest, listener);
+            }
+
+            @Override
+            public void failed(Exception e) {
+                logger.error("failed to execute pipeline [{}]", e, pipelineId);
+                listener.onFailure(e);
+            }
+        });
     }
 
-    @Override
-    protected boolean apply(String action, ActionResponse response, ActionListener listener) {
-        return true;
+    // TODO: rethink how to deal with bulk requests:
+    // This doesn't scale very well for a single bulk requests, so it would be great if a bulk requests could be broken up into several chunks so that the ingesting can be paralized
+    // on the other hand if there are many index/bulk requests then breaking up bulk requests isn't going to help much.
+    // I think the execution service should be smart enough about when it should break things up in chunks based on the ingest threadpool usage,
+    // this means that the contract of the execution service should change in order to accept multiple data instances.
```

Comment:
actually.. thought more about it. The above assumes something that I'm not sure we should assume - all pipelines created (and therefore treated) equally. Perhaps we should look at it differently:

- conceptually, it's much easier to think of things when assigning a thread per pipeline execution (so it's not on a request level but on a pipeline level)
- I imagine scenarios where there are multiple pipelines, some are more complex than others. One pipeline does very simple and fast processing, another is more complex and involved "heavier" computations (perhaps connect to 3rd party system on cache expiry of some sort). 
- If the above case applies, we should enable assigning a threadpool per pipeline. So for example, we define a default thread pool for all pipelines... but in its simplest form, perhaps a pipeline can define the thread pool it should be executed with. This way users can create dedicated TP for specific pipeline executions.

would love to get feedback from the LS team here... perhaps I'm completely off with my thinking here